crates/matrix-sdk-crypto-nodejs/src/machine.rs

```diff
@@ -0,0 +1,379 @@
+    }
+
+    #[napi]
+    pub fn mark_request_as_sent(
+        &self,
+        request_id: String,
+        request_kind: RequestKind,
+        response_body: String,
+    ) -> Result<()> {
+        let req_id = Uuid::parse_str(request_id.as_str()).expect("Failed to parse request ID");
+        let response = response_from_string(response_body.as_str());
+
+        let response: OwnedResponse = match request_kind {
+            RequestKind::KeysUpload => {
+                KeysUploadResponse::try_from_http_response(response).map(Into::into)
+            }
+            RequestKind::KeysQuery => {
+                KeysQueryResponse::try_from_http_response(response).map(Into::into)
+            }
+            RequestKind::ToDevice => {
+                ToDeviceResponse::try_from_http_response(response).map(Into::into)
+            }
+            RequestKind::KeysClaim => {
+                KeysClaimResponse::try_from_http_response(response).map(Into::into)
+            }
+            RequestKind::SignatureUpload => {
+                SignatureUploadResponse::try_from_http_response(response).map(Into::into)
+            }
+            RequestKind::KeysBackup => {
+                KeysBackupResponse::try_from_http_response(response).map(Into::into)
+            }
+            RequestKind::RoomMessage => {
+                RoomMessageResponse::try_from_http_response(response).map(Into::into)
+            }
+        }
+        .expect("Can't convert json string to response");
+
+        self.runtime
+            .block_on(self.inner.mark_request_as_sent(&req_id, &response))
+            .expect("Failed to mark request as sent");
+
+        Ok(())
+    }
+
+    #[napi]
+    pub fn receive_sync_changes(
+        &self,
+        events: String,
+        device_changes: DeviceLists,
+        key_counts: Map<String, Value>,
+        unused_fallback_keys: Option<Vec<String>>,
+    ) -> Result<String> {
+        // key_counts: Map<String, String> (cast to Map<String, i32>)
+
+        let events: ToDevice = serde_json::from_str(events.as_str())?;
+        let device_changes: RumaDeviceLists = device_changes.into();
+        let key_counts: BTreeMap<DeviceKeyAlgorithm, UInt> = key_counts
+            .into_iter()
+            .map(|(k, v)| {
+                (
+                    DeviceKeyAlgorithm::try_from(k).expect("Failed to convert key algorithm"),
+                    v.as_str()
+                        .expect("Failed to get string for number")
+                        .parse::<i32>()
+                        .unwrap()
+                        .try_into()
+                        .expect("Failed to convert to number"),
+                )
+            })
+            .collect();
+
+        let unused_fallback_keys: Option<Vec<DeviceKeyAlgorithm>> =
+            unused_fallback_keys.map(|u| u.into_iter().map(DeviceKeyAlgorithm::from).collect());
+
+        let events = self
+            .runtime
+            .block_on(self.inner.receive_sync_changes(
```

Comment:
Almost all methods here are async, you are converting them to be sync using `block_on()`. Uniffi sadly doesn't support automagical conversions betewen Rust async methods and Kotlin suspendable ones. napi-rs does, so we should use it.

```rust
let machine = self.inner.clone();
let events = self.runtime.spawn(async move {
    machine.receive_sync_changes(
        events,
        &device_changes,
        &key_counts,
        unused_fallback_keys.as_deref(),
    ).await
})
.await
.expect("Task paniced")
.expect("Failed to handle sync changes");
```

You might want to continue using `block_on()` for the constructor and perhaps some other hand picked methods but certainly not for all of these.

Reply:
Actually napi-rs seems to support tokio out of the box, so you don't need to create a runtime at all:

> NAPI-RS supports the tokio runtime by default. If you await a tokio future in async fn, NAPI-RS will execute it in the tokio runtime and convert it into a JavaScript Promise.

More info here: https://napi.rs/docs/concepts/async-fn.